fix: prevent collection mapping using invalid length property

TimothyMakkison opened the pull request:
Fix bug where array to list mappings would attempt to use an invalid `Length` property on the lowered `IReadOnlyCollection` source.

Fixes #1106

- `EnumerableToList` will always use `Count` instead of the non lowered sources `CountPropertyName`
  - I don't think this will cause issues, I did consider using `CollectionInfoBuilder` to get the new `CountPropertyName`
- Added `ArrayToListShouldImproveNullability` test
- Not sure if a similar issues exists with `BuildEnumerableToArrayMapping`, I think all calls to this method will always use `Count` for the original and lowered types. Perhaps `CountPropertyName` should be used here

## Checklist

- [x] The existing code style is followed
- [x] The commit message follows our guidelines
- [x] Performed a self-review of my code
- [x] Unit tests are added/updated

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -200,13 +201,7 @@ or CollectionType.IEnumerable
-            AddMethodName,
-            ctx.CollectionInfos!.Source.CountPropertyName
-        );
+        return new ForEachAddEnumerableMapping(sourceType, targetType, elementMapping, AddMethodName, CountPropertyName);
```
I think your indentation is correct, I think the problem also exists in `EnumerableMappingBuilder.BuildEnumerableToArrayMapping`.
Also currently we accept `Length` as count property names of custom collections, don't we? I think a custom collection type as source which only implements `IEnumerable<T>` with a `int Length` property won't work with this fix.

Another approach would probably be in 
`EnumerableMappingBuilder.BuildCollectionTypeForICollection` instead of only returning the `INamedTypeSymbol`, also return the name of the count property:
```c#
private static (ITypeSymbol Type, string? CountPropertyName) BuildCollectionTypeForICollection(MappingBuilderContext ctx, CollectionInfo info)
    {
        return info.ImplementedTypes.HasFlag(CollectionType.IReadOnlyCollection)
            ? (BuildCollectionType(ctx, CollectionType.IReadOnlyCollection, info.EnumeratedType), nameof(IReadOnlyCollection<object>.Count))
            : info.ImplementedTypes.HasFlag(CollectionType.ICollection)
                ? (BuildCollectionType(ctx, CollectionType.ICollection, info.EnumeratedType), nameof(ICollection<object>.Count))
                : (info.Type, info.CountPropertyName);
    }
```

or alternatively we could probably return an entire new `CollectionInfo` for `IReadOnlyCollection<T>` / `ICollection<T>` (I'd probably prefer this 🤔).

latonz reviewed (changes requested):
Thank you for looking into this and contributing a fix!

Doesn't a similar problem exist in `EnumerableMappingBuilder.BuildCustomTypeMapping` with the `EnsureCapacityBuilder.TryBuildEnsureCapacity(ctx)` call using the old / wrong source count property? IMO the best solution would be to create a new `CollectionInfos` instance when overwriting the source type and passing that directly to `TryBuildEnsureCapacity`.
I think also `DictionaryMappingBuilder.BuildCustomTypeMapping` has a potential similar problem... I don't think it is a real problem there now (as it is always the Count property for dictionary), but I don't think the code now is really correct.

TimothyMakkison commented:
> Another approach would probably be in
EnumerableMappingBuilder.BuildCollectionTypeForICollection instead of only returning the INamedTypeSymbol, also return the name of the count property:

> CollectionInfo for IReadOnlyCollection<T> / ICollection<T> (I'd probably prefer this 🤔).

I opted to return `CollectionInfo` from `BuildCollectionTypeForICollection`.

- Made `CollectionInfoBuilder.BuildCollectionInfo` public, passing in `SymbolAccessor` and `WellKnownTypes` is a bit of pain.
- Updated `TyrBuildEnsureCapacity` to `TryBuildEnsureCapacity(MappingBuilderContext ctx, CollectionInfo source, CollectionInfo target)`
- Updated all mappers to use the new `EnsureCapacity`
- `DictionaryMappingBuilder` uses the correct `IsCountKnown` and `CountProperty` values.

latonz reviewed (changes requested):
Thank you for the updates, one small feedback point.

TimothyMakkison commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -10,26 +10,26 @@ public static class EnsureCapacityBuilder
     private const string EnsureCapacityName = "EnsureCapacity";
     private const string TryGetNonEnumeratedCountMethodName = "TryGetNonEnumeratedCount";
 
-    public static EnsureCapacityInfo? TryBuildEnsureCapacity(MappingBuilderContext ctx)
```
Great change, I've also moved the `ctx.CollectionInfos` null check into this.
